WIP stream-layer flow classify hook proposal (#4545)
microsoft/ebpf-for-windows: docs/FlowClassifyHook.md

    - [Program Type](#program-type)
    - [Hooks](#hooks)
5. [Architecture](#architecture)
    - [Hook Integration and Flow](#hook-integration-and-flow)
    - [Stream Hook Lifecycle](#stream-hook-lifecycle)

---

## Purpose

Support an eBPF interface to support inspecting TCP stream data and then based on that allowing/blocking the connection.

The new hooks will support security and observability related solutions that require parsing TCP stream data
without incurring overhead for flows that can be ignored.

## Requirements

- Hook that allows choosing whether to classify a newly established TCP connection at the stream layer.
- Hook that receives each stream layer data segment in-order for both ingress and egress traffic.
  - 3 possible actions:
    - Allow the connection and stop being invoked for this flow.
    - Block the connection (no further invocations for this flow).
    - Allow the segment but keep getting invoked for further data segments (need more data to classify).
- Hook that supports cleanup of flows that were closed while still being classified.
- No eBPF programs should be invoked for segments on flows not needing stream layer classification.

> - After a flow is allowed/block, there should be no further stream layer eBPF program invocations for the flow.

Author: I've updated the proposal to describe the multiple program support.

One complication that adds is: which program(s) should be invoked for the `EBPF_ATTACH_TYPE_FLOW_DELETED` hook?
- All of them?
- All that are also attached to `STREAM_FLOW_CLASSIFY` but did not already return `ALLOW`/`BLOCK`?

Alternatively, this could be made a single attach type `EBPF_ATTACH_TYPE_STREAM_FLOW_CLASSIFY`, then the extension can track each program through each flow easily and invoke the same bpf program on a new flow, for each segment until classified, and then if the flow is deleted before classification.

**Combining the attach types into 1 seems the simplest**. That loses the option of attaching separate programs to each hook, but avoids needing to specify relationships between the programs and simplifies flow lifecycle management.

Thoughts?

typedef enum _flow_state
{
    FLOW_STATE_NEW,
    FLOW_STATE_ESTABLISHED,
    FLOW_STATE_DELETED,
    FLOW_STATE_INVALID,
} flow_state_t;

typedef struct _bpf_flow_classify
{
    uint32_t family; ///< IP address family.
    struct
    {
        union
        {
            uint32_t local_ip4;
            uint32_t local_ip6[4];
        }; ///< Local IP address.
        uint32_t local_port;
    }; ///< Local IP address and port stored in network byte order.
    struct
    {
        union
        {
            uint32_t remote_ip4;
            uint32_t remote_ip6[4];
        }; ///< Remote IP address.
        uint32_t remote_port;
    }; ///< Remote IP address and port stored in network byte order.
    uint8_t protocol;        ///< IP protocol.
    uint32_t compartment_id; ///< Network compartment Id.
    uint64_t interface_luid; ///< Interface LUID.
    uint8_t direction;       ///< 0 = inbound, 1 = outbound
    uint64_t flow_id;        ///< WFP flow handle
    uint32_t state;          ///< State of the flow.
    uint8_t* data_start;     ///< Pointer to start of stream segment data
    uint8_t* data_end;       ///< Pointer to end of stream segment data
} bpf_flow_classify_t;
```

> ### Hooks

Author: Added attach type details.

1. **new_flow_classify**
   - **Invocation:** Invoked when a new TCP flow is established (at the WFP flow established layer).
   - **Purpose:** Allows the eBPF program to make an initial classification decision based on connection metadata (addresses, ports, process, etc.) before any stream data is seen.
   - **Return values:**
     - `FLOW_CLASSIFY_ALLOW`: The flow is permitted and will not be further inspected at the stream layer; no additional hook invocations for this flow.
     - `FLOW_CLASSIFY_NEED_MORE_DATA`: The flow requires inspection at the stream layer; subsequent TCP segments will trigger `stream_flow_classify` invocations.
   - **Notes:** This avoids unnecessary stream processing for flows that do not require data inspection at the stream layer.

> 2. **stream_flow_classify**

Reviewer: ```suggestion
2. **EBPF_ATTACH_TYPE_STREAM_FLOW_CLASSIFY**
```

   - **Invocation:** Called for each TCP segment on flows that were marked as needing more data by `new_flow_classify`.
   - **Purpose:** Allows the eBPF program to inspect stream data and allow/block the connection, or request more data if classification is not yet possible.
   - **Return values:**
     - `FLOW_CLASSIFY_ALLOW`: The flow is permitted; no further stream inspection for this flow.
     - `FLOW_CLASSIFY_BLOCK`: The flow is blocked; all further data is dropped (using WFP ABSORB), and the connection will be terminated or time out.
     - `FLOW_CLASSIFY_NEED_MORE_DATA`: The current segment is allowed, but the program will be invoked again for subsequent segments until a final decision is made.
   - **Notes:** This is only invoked for each flow segment until allow/block are returned.

> 3. **flow_deleted**

Reviewer: ```suggestion
3. **EBPF_ATTACH_TYPE_FLOW_DELETED**
```

   - **Invocation:** Called when a flow is deleted (e.g., connection teardown) and the last action for the flow was `FLOW_CLASSIFY_NEED_MORE_DATA` (i.e., the program never returned a final allow/block decision).
   - **Purpose:** Allows the eBPF program to perform cleanup of any per-flow state (e.g., in maps) that was maintained during inspection.
   - **Return value:**
     - Ignored
   - **Notes:** This hook is only invoked for flows that were still being actively classified and not for those that were allowed immediately.

```c
/*
 * @brief Select flows for stream layer classification
 *
 * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
 *
 * Attach type(s):
 * \ref EBPF_ATTACH_TYPE_NEW_FLOW_CLASSIFY
 *
 * @param[in] context \ref bpf_flow_classify_t

>  * @return ALLOW to ignore(allow) this flow, NEED_MORE_DATA for stream layer classification.

Reviewer: ```suggestion
 * @retval ALLOW The flow should be allowed.
 * @retval NEED_MORE_DATA More data is needed to determine whether the flow should be allowed.
```

Use retval not return, to document specific return values.
Also, why is BLOCK not allowed here?

 */
typedef flow_classify_action_t
new_flow_classify_hook_t(bpf_flow_classify_t* context);

/*
 * @brief Handle flow classification (stream inspection).
 *
 * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
 *
 * Attach type(s):
 * \ref EBPF_ATTACH_TYPE_STREAM_FLOW_CLASSIFY
 *
 * @param[in] context \ref bpf_flow_classify_t

>  * @return classification decision (allow, block, or need more data to decide).

Reviewer: Use retval with each value.

 */
typedef flow_classify_action_t
stream_flow_classify_hook_t(bpf_flow_classify_t* context);

/*
 * @brief Handle deletion of a flow that is still being classified
 *
 * Called for flows where the last return value was NEED_MORE_DATA
 *
 * Program type: \ref EBPF_PROGRAM_TYPE_FLOW_CLASSIFY
 *
 * Attach type(s):
 * \ref EBPF_ATTACH_TYPE_FLOW_DELETED
 *
 * @param[in] context \ref bpf_flow_classify_t
 * @return classification decision (allow, block, or need more data to decide).
 */
typedef flow_classify_action_t
flow_deleted_hook_t(bpf_flow_classify_t* context);
```

> ## Architecture

Reviewer 2: Consider adding more explicit details about what the hook is doing at the WFP layer. A few questions that I have:
- I see you have proposed a BPF need_more_data value, but does that correspond to the FWPS_STREAM_ACTION_NEED_MORE_DATA  value?
- The stream inspection callout has references to things like countBytesEnforced - how are we setting this value in the hook?
- Are we doing inline stream inspection only?
- We must ensure that we are following the waterfall model in our API usage. It may be good to consider adding some details about how we are doing it (as this is the crux of a stream WFP callout, and places that can cause kernel crashes if done incorrectly).

### Hook Integration and Flow

The extension uses the Windows Filtering Platform (WFP) to register callouts at both the flow established and stream layers.

1. When a new TCP flow is established, the extension initializes a per-flow context and invokes the `new_flow_classify` hook.

> 2. The program can return ALLOW/NEED_MORE_DATA to ignore the flow or classify the flow at the stream layer.

Reviewer: "The program" implies there can only be one such program.  I think this should instead say "Each program" and allow multiple programs.

>     b. If no stream-layer inspection is needed, the bpf program context is freed immediately.

Reviewer 2: nit: b and a are flipped

- When returning NEED_MORE_DATA, the current data segment is allowed (but the hook will keep getting invoked).
- Multiple eBPF programs can be invoked for a single stream, but the decision is finalized based on the first conclusive return code.
- No stream mutation is supported through these hooks (only inspection and allow/block).

### WFP Implementation Details

The stream inspection extension integrates with the Windows Filtering Platform (WFP)
at the flow_established layer to indicate flows for classification and at the stream layer to
classify flows.

Currently only inline stream classification will be supported.

#### WFP Action Mapping

The hook implementation maps eBPF program return values to specific WFP actions:

- **FLOW_CLASSIFY_ALLOW**: Returns `FWP_ACTION_PERMIT` and deletes the WFP flow context. This allows the current segment and terminates further classification for this flow by removing the flow context.

- **FLOW_CLASSIFY_BLOCK**: Returns `FWP_ACTION_BLOCK` with the `FWPS_CLASSIFY_OUT_FLAG_ABSORB` flag set. This blocks the current segment and terminates the connection by dropping all subsequent packets.

> - **FLOW_CLASSIFY_NEED_MORE_DATA**: Returns `FWP_ACTION_PERMIT`. This allows the current segment to proceed while maintaining the flow context for continued classification of subsequent segments.

Reviewer 2: Is there a reason why we do this, as opposed to using the FWPS_STREAM_ACTION_NEED_MORE_DATA?

- Data is allowed through when `FLOW_CLASSIFY_NEED_MORE_DATA` is returned.
  - The allow allow/block decision is at the flow level (not segment level).
  - **Note**: The `countBytesEnforced` field is not currently used in this implementation as the hook focuses on connection-level allow/block decisions rather than byte-level enforcement.

#### Stream Data Handling

The hook provides direct access to stream segment data through the context's `data_start` and `data_end` pointers:

- **Contiguous Data**: When possible, pointers reference the original NET_BUFFER data directly for zero-copy access
- **Non-contiguous Data**: For fragmented NET_BUFFERs, data is copied into a temporary buffer using `FwpsCopyStreamDataToBuffer0` to provide contiguous access to eBPF programs
- **Data Lifetime**: Stream data pointers are only valid during the program invocation and must not be accessed after the program returns

#### Resource Management

The implementation manages WFP resources efficiently:

- **Conditional Callouts**: Stream layer callouts use the `FWP_CALLOUT_FLAG_CONDITIONAL_ON_FLOW` flag to ensure they are only invoked for flows requiring classification

> - **Context Cleanup**: Flow contexts are automatically cleaned up when flows are allowed, blocked, or naturally terminated

Reviewer 2: For both context cleanup and memory management - can you update the text slightly, to mention that the stream extension does the following?

Alternatively, if it is WFP that does something, please call that out here too. Sometimes, WFP callouts hand memory off to WFP platform, which is responsible for cleaning it up later, so it is good to be precise here about what we are responsible for/doing w.r.t. memory.)
